juju/juju
# Add support for region-config in cloud.yaml

This adds initial support for region-config as part of the coming
model-config-tree change. 

@wallyworld @axw 

QA Steps:
Build create a yaml file something like: 
```
awstest:
  type: ec2
  regions:
    us-west-1:
      endpoint: https://us-west-1.aws.amazon.com/v1.2/
    us-east-1:
      endpoint: https://us-east-1.aws.amazon.com/v1.2/
  config:
    bootstrap-timeout: 900
      set-numa-control-policy: true
  region-config:
    us-east-1:
      apt-mirror: https://foo-east
    us-west-1:
      apt-mirror: https://foo-west
```
Import it:
`$ juju add-cloud awstest aws-region-config-tree.yaml`

Check we have it:
```
juju show-cloud local:awstest
defined: local
type: ec2
regions:
  us-west-1:
    endpoint: https://us-west-1.aws.amazon.com/v1.2/
  us-east-1:
    endpoint: https://us-east-1.aws.amazon.com/v1.2/
config:
  bootstrap-timeout: 900
  set-numa-control-policy: true
region-config:
  us-east-1:
    apt-mirror: https://foo-east
  us-west-1:
    apt-mirror: https://foo-west
```
Bootstrap to aws:

`juju bootstrap region-config-test awstest`

Make some coffee.
Make a sandwich.

Eventually you should be able to log into the controller and see that region-config is retained in:
`/var/lib/juju/bootstrap-params`

Add a model
`juju add-model foobar`

And looking in the database on the controller it is in the `globalSettings` collection:
```
juju:PRIMARY> db.globalSettings.find()
{ "_id" : "controllerInheritedSettings", "model-uuid" : "", "settings" : {  }, "version" : NumberLong(0), "txn-revno" : NumberLong(2), "txn-queue" : [ "579be0353cbb4221674450b4_d20babcc" ] }
{ "_id" : "awstest#us-east-1", "model-uuid" : "", "settings" : { "apt-mirror" : "https://foo-east" }, "version" : NumberLong(0), "txn-revno" : NumberLong(2), "txn-queue" : [ "579be0353cbb4221674450b4_d20babcc" ] }
{ "_id" : "awstest#us-west-1", "model-uuid" : "", "settings" : { "apt-mirror" : "https://foo-west" }, "version" : NumberLong(0), "txn-revno" : NumberLong(2), "txn-queue" : [ "579be0353cbb4221674450b4_d20babcc" ] }
```




(Review request: http://reviews.vapour.ws/r/5339/)
